text/0000-char-uax-31.md

```diff
@@ -0,0 +1,103 @@
+- Feature Name: `char_uax_31`
+- Start Date: 2019-24-04
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add functions to the standard library for testing a `char` against [UAX TR31](https://unicode.org/reports/tr31/) ("Unicode Annex 31")
+`Pattern_White_Space`, `Pattern_Syntax`, `XID_Start`, `ID_Nonstart`, and `XID_Continue` (the XID ones are already in the standard
+library, but are unstable; this RFC proposes to stablize them).
+
+# Motivation
+[motivation]: #motivation
+
+As a systems language, Rust is heavily used for parsing.
+As a progressive, forward-thinking language that accepts anyone,
+Rust supports Unicode and makes the definitive string types UTF-8.
+At the intersection of these needs sits *UAX #31: Unicode Identifier and Pattern Syntax* ("Annex 31"),
+a standardized set of code point categories for defining computer language syntax.
+
+This is being used in production Rust code already.
+Rust's own compiler already has functions to check against Annex 31 code point categories in the lexer,
```

Comment:
This is for the unstable, old, non_ascii_idents feature, which [we have already RFCd to change](https://github.com/rust-lang/rfcs/blob/master/text/2457-non-ascii-idents.md). The change will still need _some_ function like this, but it may need tailoring for bidi characters. This is true for other implementors too, the XID functions often need tailoring.

They're not super hard to tailor, though, so we could still expose these functions to match spec and let those tailoring just suffix the calls with `|| ch == ... && ch != ...`

Reply:
The XID part is currently unstable. But stable Rust does respect `Pattern_White_Space`, so there's already committed-to Annex 31-based syntax.

Reply:
As a concrete case of tailoring, Rust actually uses `XID_Start | '_'` and not a plain `XID_Start`. I've actually got bitten by that when implementing lexer for IntelliJ Rust a while ago :)